cpp/source/deal.II/mp_values.templates.h

```diff
@@ -6,82 +6,273 @@
-                                         std::vector<double> &values) const
-
+std::unique_ptr<MPValues<dim, spacedim>>
+_build_material_properties(std::string const &material_name,
+                           MPValuesParameters<dim, spacedim> const &params)
 {
-  dealii::types::material_id const material_id = cell->material_id();
-  auto got = (this->materials).find(material_id);
-  if (got == (this->materials).end())
+  boost::property_tree::ptree const &database = *params.database;
+  boost::property_tree::ptree const &material_database =
+      database.get_child(material_name);
+  std::string const type = material_database.get<std::string>("type");
+  if (type.compare("porous_electrode") == 0)
   {
-    std::cerr << "failed to find\n";
-    std::cerr << "id=" << material_id
-              << "  int=" << static_cast<int>(material_id)
-              << "  to_string=" << std::to_string(material_id) << "\n";
-    std::cerr << "in\n";
-    for (auto x : this->materials)
-      std::cerr << "id=" << x.first << "  int=" << static_cast<int>(x.first)
-                << "  to_string=" << std::to_string(x.first) << "\n";
-    throw std::runtime_error("Invalid material id " +
-                             std::to_string(material_id));
+    return std::make_unique<PorousElectrodeMPValues<dim, spacedim>>(
+        material_name, params);
+  }
+  else if (type.compare("permeable_membrane") == 0)
+  {
+    std::string const matrix_phase =
+        database.get<std::string>(material_name + "." + "matrix_phase");
+
+    auto dummy_database =
+        std::make_shared<boost::property_tree::ptree>(*params.database);
+    dummy_database->put(matrix_phase + "." + "differential_capacitance", 0.0);
+    dummy_database->put(matrix_phase + "." + "exchange_current_density", 0.0);
+    dummy_database->put(matrix_phase + "." + "electrical_resistivity",
+                        std::numeric_limits<double>::max());
+    MPValuesParameters<dim, spacedim> dummy_params = params;
+    dummy_params.database = dummy_database;
+    return std::make_unique<PorousElectrodeMPValues<dim, spacedim>>(
+        material_name, dummy_params);
+  }
+  else if (type.compare("current_collector") == 0)
+  {
+    return std::make_unique<MetalFoilMPValues<dim, spacedim>>(material_name,
+                                                              params);
+  }
+  else
+  {
+    throw std::runtime_error("Invalid material type " + type);
   }
-  auto material = got->second;
-  material->get_values(key, cell, values);
 }
 
 template <int dim, int spacedim>
-void MPValues<dim, spacedim>::get_values(
-    std::string const &key, active_cell_iterator const &cell,
-    std::vector<dealii::Tensor<1, spacedim>> &values) const
+InhomogeneousSuperCapacitorMPValues<dim, spacedim>::
+    InhomogeneousSuperCapacitorMPValues(
+        MPValuesParameters<dim, spacedim> const &params)
+{
+  // Construct a random number generator and use the MPI rank as a seed.
```

Comment:
That's fine for now. But we need to serialize MPValues if we want to do a restart with a different number of procs.